File: Runtime/Plugins/Android/DeferredDeeplinkParametersListenerProxy.java
package io.appmetrica.analytics.plugin.unity;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import org.jetbrains.annotations.NotNull;
import org.json.JSONException;
import org.json.JSONObject;
import java.util.Map;
import io.appmetrica.analytics.DeferredDeeplinkListener;
import io.appmetrica.analytics.DeferredDeeplinkParametersListener;
import io.appmetrica.analytics.StartupParamsCallback;
import io.appmetrica.analytics.StartupParamsItem;
import io.appmetrica.analytics.StartupParamsItemStatus;

public interface DeferredDeeplinkParametersListenerProxy {
    void onParametersLoaded(@NotNull String parameters);

    void onError(@NotNull String error, @NotNull String referrer);

    class Delegate implements DeferredDeeplinkParametersListener {
        @NonNull
        private final DeferredDeeplinkParametersListenerProxy proxy;

        public Delegate(@NonNull DeferredDeeplinkParametersListenerProxy proxy) {
            this.proxy = proxy;
        }

        @Override
        public void onParametersLoaded(@NonNull Map<String, String> map) {
            proxy.onParametersLoaded(new JSONObject(map).toString());
        }

        @Override
        public void onError(@NonNull Error error, @NonNull String referrer) {
            proxy.onError(getErrorStr(error), referrer);
        }

        @NonNull
        private static String getErrorStr(@NonNull Error error) {
            switch (error) {
                case NO_REFERRER:
                    return "NoReferrer";
                case NOT_A_FIRST_LAUNCH:
                    return "NotAFirstLaunch";
                case PARSE_ERROR:
                    return "ParseError";
                default:
                    return "Unknown";
            }
        }
    }
}

File: Runtime/Native/Android/Proxy/DeferredDeeplinkParametersListenerProxy.cs
#if UNITY_ANDROID
using Io.AppMetrica.Internal;
using JetBrains.Annotations;
using System.Linq;
using UnityEngine;

namespace Io.AppMetrica.Native.Android.Proxy {
    internal class DeferredDeeplinkParametersListenerProxy : AndroidJavaProxy {
        [NotNull]
        private readonly DeferredDeeplinkParameters.ParametersDelegate _onParametersLoadedDelegate;
        [CanBeNull]
        private readonly DeferredDeeplinkParameters.ErrorDelegate _onErrorDelegate;

        public DeferredDeeplinkParametersListenerProxy(
            [NotNull] DeferredDeeplinkParameters.ParametersDelegate onParametersLoadedDelegate,
            [CanBeNull] DeferredDeeplinkParameters.ErrorDelegate onErrorDelegate
        ) : base("io.appmetrica.analytics.plugin.unity.DeferredDeeplinkParametersListenerProxy") {
            _onParametersLoadedDelegate = onParametersLoadedDelegate;
            _onErrorDelegate = onErrorDelegate;
        }

        // ReSharper disable once InconsistentNaming
        public void onParametersLoaded([NotNull] string parameters) {
            _onParametersLoadedDelegate(
                JSONDecoder.Decode(parameters).ObjectValue.ToDictionary(
                    keySelector: it => it.Key,
                    elementSelector: it => it.Value.StringValue
                )
            );
        }

        // ReSharper disable once InconsistentNaming
        public void onError([NotNull] string error, [NotNull] string referrer) {
            _onErrorDelegate?.Invoke(GetError(error), referrer);
        }

        private static DeferredDeeplinkParameters.Error GetError([NotNull] string str) {
            switch (str) {
                case "NoReferrer":
                    return DeferredDeeplinkParameters.Error.NoReferrer;
                case "NotAFirstLaunch":
                    return DeferredDeeplinkParameters.Error.NotAFirstLaunch;
                case "ParseError":
                    return DeferredDeeplinkParameters.Error.ParseError;
                default:
                    return DeferredDeeplinkParameters.Error.Unknown;
            }
        }
    }
}
#endif
